pkg/controller/virtualservice.go

```diff
@@ -288,6 +296,30 @@ func vserviceNeedsUpdate(desired *appmeshv1beta1.VirtualService, target *aws.Vir
 	return false
 }
 
+func vrouterNeedsUpdate(desired *appmeshv1beta1.VirtualRouter, target *aws.VirtualRouter) bool {
+	if desired.Name != target.Name() {
+		return true
+	}
+
+	if len(desired.Listeners) != len(target.Data.Spec.Listeners) {
+		return true
+	}
+
+	if len(desired.Listeners) > 0 {
+		//there should be only one listener
```

Comment:
Somewhat tricky logic to read, but it makes sense. I was just considering the options:
```
+------------------------+---------------------------------+-------------------+
| len(desired.Listeners) | len(target.Data.Spec.Listeners) | routerNeedsUpdate |
+------------------------+---------------------------------+-------------------+
|           0            |               0                 |       false       |
|                        |                                 |                   |
|           1            |               0                 |       true        |
|                        |                                 |                   |
|           0            |               1                 |       true        |
|                        |                                 |                   |
|           1            |               1                 |       true        |
|                        |                                 |                   |
|           2            |               1                 |       true        |
+------------------------+---------------------------------+-------------------+
```
I assume that `target.Data.Spec.Listeners` contains at most `1` listener?

Reply:
yeah I can change this to a loop and assume order is preserved or be defensive and error if > 1. either ways it is still tricky :)